haas/config.py

```diff
@@ -17,19 +17,93 @@
 This module handles loading of the haas.cfg file, and querying the options
 therein. the `cfg` attribute is an instance of `ConfigParser.RawConfigParser`.
 Once `load` has been called, it will be ready to use.
+
+Logging is initialized here.
 """
 
-import ConfigParser
+import ConfigParser, logging, os
+
+# If not specified otherwise, this is the default config file read
+DEFAULT_CONFIG_FILENAME='haas.cfg'
 
+# cfg is used extensively by other parts of the program and represents a
+# "public" interface
 cfg = ConfigParser.RawConfigParser()
 
-def load(filename='haas.cfg'):
-    """Load the configuration from the file 'haas.cfg' in the current directory.
+# Keep track of the loaded config file.
+cfgFile = None
+
+def reset():
+    """
+    Reset the cfg global. Return the previous config file. Used for unit
+    tests.
+    If you want to use this to reload the config file in normal operation,
+    users of 'cfg' will need to use an accessor function instead of cfg
+    directly since they could cache a stale copy.
+    """
+    global cfg, cfgFile
+    cfg = ConfigParser.RawConfigParser()
+
+    temp = cfgFile
+    cfgFile = None
+    return temp
+
+import os.path
+def load(filename=None, requireConfigFile=True):
+    """Load the configuration from exactly one file (or file-like object)
+    indicated by (in order):
+        1) The argument passed (most likely a command line argument or the
+        testing framework)
+        2) The environmental variable "HAAS_CONFIG"
+        3) 'haas.cfg' in the current directory
 
     This must be called once at program startup; no configuration options will
-    be available until then.
+    be available until then. Once called, further calls to load() will silently
+    be ignored.
 
-    If the configuration file is not available, this function will simply load
-    an empty configuration (i.e. one with no options).
+    If the configuration file is not available, throw an exception unless
+    requireConfigFile was set to False. We generally (for server configs) need
```

Comment:
re: our conversation today; the few things that need more than just the endpoint really *are* special; I actually like the idea of just listing them explicitly.